Mathlib/Tactic/DeriveInfinite.lean

```diff
@@ -0,0 +1,328 @@
+    let val ← whnf val
+    if let .const ctor _ := val.getAppFn then
+      let ctorInfo ← getConstInfoCtor ctor
+      return some ctorInfo.name
+    else
+      return none
+  catch _ =>
+    return none
+
+/--
+Try to construct a proof for `Infinite α` by forming an injection from a `α`-valued field to the
+subset of `α` created with this constructor, and then using `Infinite.of_injective_to_set`.
+To prove that this subset is proper, we show that the `default` element of `α` is not headed by
+this constructor (if `default` is headed by this constructor, then this function fails).
+-/
+private def mkProofSelf (ctorInfo : ConstructorVal) (xs : Array Expr) (self : Expr) (defaultCtor : Name) : TermElabM (Option Term) := do
+  if ctorInfo.name == defaultCtor then
+    return none
+
+  let mut selfField := none
+  for i in [:ctorInfo.numFields] do
+    let x := xs[ctorInfo.numParams + i]!
+    let xTy ← inferType x
+    let isSelf ← withNewMCtxDepth (isDefEq self xTy)
+    let nonempty ← trySynthInstance (← mkAppM ``Nonempty #[xTy]) <&> LOption.toOption
+
+    trace[Elab.Deriving.infinite] "{ctorInfo.name} field {i}: isSelf={isSelf} nonempty={nonempty} value={toString x} value={x} type={toString xTy} type={xTy}"
+
+    match selfField, isSelf, nonempty with
+    | _, false, none =>
+      trace[Elab.Deriving.infinite] "{ctorInfo.name} field {i}: not Nonempty or self; giving up on this constructor"
+      return none -- field is neither self nor Nonempty
+    | none, true, _ => selfField := some i -- first self field found
+    | _, true, none => selfField := some i -- this field is only self, not Nonempty
+    | _, _, _ => pure ()
+
+  let some j := selfField
+    | return none
+  let (obj, holeArg, addNonemptyElim) ← mkObj ctorInfo j
+  let noConfusion := mkCIdent <| ctorInfo.induct.str "noConfusion"
+  let noConfusionMotive ← mkNoConfusionMotive ctorInfo xs self j
+  let val ←
+    `(let S := {x | ∃ $holeArg:ident, $obj = x}
+      let notUniv : S ≠ $(mkCIdent ``Set.univ) := fun h =>
+        (h ▸ show $(mkCIdent ``default) ∉ S from fun ⟨_, h₂⟩ => $noConfusion h₂) $(mkCIdent ``trivial)
+      let f : _ → S := fun $holeArg:ident => ⟨$obj, $holeArg, rfl⟩
+      let inj : $(mkCIdent ``Function.Injective) f := fun _ _ h =>
+        $(mkCIdent ``Subtype.noConfusion) h fun h₂ => $noConfusion h₂ $noConfusionMotive
+      $(mkCIdent ``Infinite.of_injective_to_set) notUniv inj)
+  addNonemptyElim val
+
+/--
+The `derive_infinite% α` term elaborator attempts to derive an instance of `Infinite α`.
+`α` could be `_` if the expected type is clear from context.
+-/
+syntax (name := derive_infinite) "derive_infinite% " term : term
+
+/-- Unifies `Infinite type` with `expectedType?` and returns `type`. -/
+private def matchTypes (type : Term) (expectedType? : Option Expr) : TermElabM (Expr × InductiveVal) := do
+  let type ← Term.elabType type
+  if let some expectedType := expectedType? then
+    let infType ← mkAppM ``Infinite #[type]
```

Comment:
In elaborator code, you want to avoid `mkAppM` because it has a check that there are no metavariables in the result (and you're wanting this code to be able to solve for some in the `isDefEq` below). This is why [this line](https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Tactic/ProxyType.lean#L239) uses `Term.elabType`.

You can use ``Term.elabType (← `(Infinite $(← Term.exprToSyntax type)))``.

Reply:
This is news to  me, I think there will be plenty of violators because it is almost never the case that elaborated exprs are metavar-free under normal elaboration conditions, even in the best case, unless the tactic has control over the whole expression generation process (as in a derive handler).

Reply:
I took a look at the code and I don't think this assertion is true @kmill . It has a check for no *assignable* mvars, but it is within a `withNewMCtxDepth` wrapper which means that most mvars will not be assignable, except for those created during the instantiation of the constant itself. That is to say, using it to construct an `Infinite A` should be fine, and if `Infinite` had e.g. a typeclass argument then that typeclass argument would need to be solved on the spot but any mvars inside `A` would not be.

Reply:
@digama0 Huh, I could have sworn that I ran into the error, but I seem to be mistaken. Thanks for the correction.

I think I confused this with a separate issue, which is that the presence of `withNewMCtxDepth` means that you can't use `mkAppM` to unify metavariables. That does not matter here.

Reply:
Thanks for the discussion here. I checked over the existing usages of `mkAppM` and everything looked fine.